### PR TITLE
Add Random Experiment

### LINKED ISSUE #51: Search The Pipeline Architecture Space
We would like to add support for doing a random or genetic algorithm approach (see TPOT for inspiration) for searching the pipeline architecture space, to improve the variety of the DAG structures in our meta-dataset.

### PR DESCRIPTION
**Pull request by epeters3**
Closes #51.

When reviewing this PR, I would recommend looking at the changes for each commit one at a time.

- Allows `EZPipeline` instance to track custom refs
- Adds more working pre-processors and classifiers and removes not working ones
- Fixes bugs related to changing default hyperparameters
- Changes data's flow through pipeline primitives
- Allows mapping pipeline arguments to custom data reference
- Unifies argument names and adds "subscribe to CLI args" paradigm
- Adds CLI arguments used by random experiment
- Adds random experiment

### CODE REVIEW

**Comment by epeters3** [comment]
It looks like the above Travis CI Branch build failed because of a test that fails sometimes with low probability, even though the test still works. So I think we can consider the builds passing for this PR.

File: experimenter/experiments/random.py [hunk 1]
```diff
@@ -1,22 +1,346 @@
...
+                # as inputs, we know they're not terminal nodes anymore.
+                terminal_node_data_refs -= input_refs
+
+                concat_result_ref = self._concatenate_inputs(structure, input_refs, concat_cache)
+                # Add the do nothing primitive as a placeholder. `self._sample_pipeline` will fill
+                # in the primitives later.
+                add_pipeline_step(
+                    structure,
+                    'd3m.primitives.data_preprocessing.do_nothing.DSBOX',
+                    concat_result_ref
```
Comment (orionw) [inline experimenter/experiments/random.py:159]: I like the `no_nothing` primitive here!

**Review by orionw — APPROVED** [review]
Very readable code!  Looks great - as does the command line args.  They're starting to get large so I think this is a good way to handle it.

**Comment by epeters3** [comment]
I’m so glad it was readable for you, thank you so much for reviewing it,
I’m sorry it was so large.

**Comment by bjschoenfeld** [comment]
> It looks like the above Travis CI Branch build failed because of a test that fails sometimes with low probability, even though the test still works. So I think we can consider the builds passing for this PR.

Can we make the test deterministic, so that it passes every time?

**Comment by orionw** [comment]
@bjschoenfeld it was supposed to test randomness - we have other tests that do it without the slight probability of failure.  I'm for just taking the test out - the other tests cover it already.  This test was supposed to be non-deterministic, back when I made it a long time ago.

**Pull request merged**